### ISSUE TITLE
Multiple Drupal VMs on a single MacBook - Shared software versions and IP addresses

### ISSUE THREAD

**Issue by captaindav** [geerlingguy/drupal-vm#1269]
### Issue Type
 - Bug Report / Support Request

### Your Environment

```
Vagrant 1.9.3
VirtualBox 5.1.18r114002
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```

### Your OS

 - **macOS** (version)

### Summary

Multiple Drupal VMs on a single MacBook - Shared software versions and IP addresses
 
I am trying to find out the best practices for having multiple Drupal VM's (often as BLT managed projects) on a single laptop. 

Here is the scenario:

On laptop #1, for project #1, a VM was created almost one year ago, using at that time, the current versions of VirtualBox, Vagrant, Ansible, (this project does not use BLT).  The software on this machine is now out of date, e.g. Vagrant version 1.8.7, Drupal VM base box 1.1.2, old version of Guest Additions, etc.

On laptop #2, for project #2, a VM was created last week using BLT. That laptop has the latest versions of VirtualBox, Vagrant, Drupal VM 4.4.0, Drupal VM base box 1.2, Ansible, Guest Additions, and BLT 8.7.0-beta1.

Now the developer assigned laptop #1 needs to work on both project #1 and project #2.

Questions:

1. Since project #2’s VM needs to be created with “blt vm” on laptop #1 using the latest version of BLT and Drupal VM included with the project, wouldn't upgrading laptop #1's Virtual Box, Vagrant, and Drupal VM base box, Guest Additions, and Ansible to the latest versions be necessary?  But wouldn’t that break project #1’s VM, because it is using an older Drupal VM base box that needs an older version of Guest Additions, and the older base box from which it was created may require an older Version of Vagrant?

2. Is there a way to avoid updating project #1's Drupal VM’s base box, while providing the latest base box version to project #2?  That is, can two Drupal VMs on the same laptop be using different versions of the DVM base box?  And would the capability for either of them to be rebuilt using “vagrant destroy/vagrant up” or “blt vm:nuke/blt vm” be available?  Or is it such that a given installation of Vagrant can only have one version of a base box from which VM’s are created?  Or perhaps there is a way in config.yml to specify not only the base box, but the base box version as well? E.g. vagrant_box_version: 1.1.2?  Kind of like a Composer lock file, so that if the base box is updated with “vagrant box update” and the VM is re-installed, the new VM still has a 1.1.2 base box.

3. If a project's base box is updated (vagrant box update), then can the project's config.yml and Vagrantfile's be re-created from the latest versions of these files found in Drupal VM or BLT? Of course, all custom settings would need to be moved from the old config.yml files to the new versions.  The instructions at https://github.com/geerlingguy/drupal-vm#updating-drupal-vm suggest reading the release notes and manually editing the existing config.yml file. But wouldn’t it be cleaner to do a diff of the old and new config.yml files, and then manually update the new version with any custom settings present in the old version? 

4. Some versions of Drupal VM are marked as "breaking", what exactly does that mean?

5.  Could it be, that given the complexity of the version inter-dependencies (e.g. Vagrant version dependent on VirtualBox version, Drupal VM box version dependent on Guest Additions version, and Ansible version etc) the best practice for having multiple Drupal VM's (or multiple BLT managed projects that use Drupal VM) is to simply require that all the projects be able to use the latest software versions?  It seems that is the most practical, since once an error is found in one of the VM's that can be fixed by updating, then the most practical way of fixing the issue is to update the entire Vagrant family of software (Virtual Box, Vagrant, plugins, Ansible, base boxes, etc.).  And thus the need for all projects to be able to use the latest software versions.

5. I see that the IP address is now set to 192.168.25.8 in the default (BLT) config.yml template.  If using host updater and auto network, can multiple VMs be run (not simultaneously – but rather one at a time) on the same OSX machine?  What was the reason for changing from to 25.8 from the old default setting of 192.168.88.88?.  It seems that running multiple VMs with the same IP causes confusion related to the “Known Hosts” .ssh file, is there a way to prevent those errors (other than deleting/editing known hosts)?

Thanks for your help!

**Comment by geerlingguy** [comment]
Eek, that's a lot of questions—I'll try to quickly answer them:

1. I have about 35 instances of Drupal VM I actively use (3 for BLT projects, the rest for other projects)—the best answer I have is to just always keep Vagrant, VirtualBox and Ansible up to date. With rare exception, _newer_ versions will always work better than old.

And IMO, you should be willing to `vagrant destroy -f` any VM at any time, since you should have all the config in your codebase, and you can just `vagrant up` and get back to the full local environment again whenever. (I probably destroy and recreate 5-10 VMs on a given day since I have to jump around projects.)

2. I think if you update a box with Vagrant, it will keep around the old one unless and until you run `vagrant box remove 'boxname'`, so as long as you don't `destroy` an old VM, it will stay put just fine. However, once you `destroy` that VM, if you have a newer box version, it will be used for the next `vagrant up` instead of the older box (you can't pin box versions to individual Vagrantfiles).

3. Whenever I do an update, I read through the release notes, but also do a quick `git diff` on any files BLT changed, then compare the extra settings I've changed to upstream variables/changes in Drupal VM. It's really tedious if you're on a _really_ old version of Drupal VM and upgrade to latest (same goes with pretty much any software... that's technical debt for you :).

But again, base boxes are basically just updated versions of the underlying OS... updating a base box is often identical to running `apt-get upgrade -y` inside an old VM. The _only_ other thing that changes is Guest Additions versions, but if you have the `vagrant-vbguest` plugin installed, you can technically use any version of VirtualBox with no ill effect (I know a few people still using VirtualBox 5.0.x, for reasons I know not, but you can do this and get it working...).

4. Breaking == there are default.config.yml variables that must be changed to ensure your configuration works identically. For example, when I cleaned up the Drupal deployment/Composer management components, some variables changed like `composer_install` to `drupal_composer_install`. I _usually_ have a shim that allows backwards compatibility (and in that case I do), but not always.

5. Yes. Yes yes yes.

Just like with Drupal itself, if you ever stop upgrading anything, you're in for a nightmarish level of technical debt. Remember what happened when people built Drupal 7 sites and left them at like 7.2 or 7.8 or whatever? Drupaggeddon quickly ruined all of those sites. There are and will be more security (and other) issues with Vagrant, VirtualBox, Ansible, macOS, Windows, etc. The only way to combat it is to stay up to date.

6. I often have 10-20 VMs running locally—there are two ways of making this happen:

  1. Make sure every VM has a unique `vagrant_ip` (I do this for every project I work on anyways)—the BLT default is just a default, it's meant to be changed.
  2. Use the `vagrant-auto_network` plugin and set `vagrant_ip` to `0.0.0.0`

Note that you can set these things in your own `local.config.yml` so that you can set whatever you want and not affect the rest of your team (if you're working on a shared project).

Basically, the takeaways:

  - Always factor in code churn and technical debt in a project—maintenance isn't a nice-to-have, it should be a top-line feature. Upgrading software, modules, libraries, etc. should be routine and occur often.
  - Use a unique IP for each VM, or use autonetwork.
  - Use `local.config.yml` to override things just for yourself.

**Comment by captaindav** [comment]
Thanks so much for your quick and clear response!  

Sorry my question was so long, I have been working on a script that automates onboarding new developers for projects managed by BLT, and built up these questions over several weeks.  The goal of the script is to create a complete, functioning local development environment in just one command, e.g. run ./onboard.sh.  I am still fine-tuning the script, hope to post it to the BLT project soon for others to use, or perhaps work to make it a BLT task, e.g. "blt local:onboard"

**Comment by geerlingguy** [comment]
@captaindav - No problem! If you have any more specific questions, please feel free to post as additional questions. And definitely at least add an issue about this on the BLT project, when you feel ready. I think there are other teams that would benefit from knowing about that strategy, even if the work didn't get merged directly into BLT.

**Comment by captaindav** [comment]
@geerlingguy - I added an issue for the BLT New Project and Onboarding scripts  captaindav/blt-example#1269, would be interested to hear if you think the approach makes sense or if they run on your machine. #1265

**Comment by i-bajrai** [comment]
@geerlingguy Question 6 is of interest to me, how do you keep track of all the IP's you have used?
